### PR TITLE
Gc.Memprof.start: take a record instead of 5 optional parameters

### PR DESCRIPTION
**Pull request by gasche**
The Gc.Memprof module provides a low-level API, that will hopefully be
paired with user libraries that provide high-level instrumentation
choices.

A natural question is: how are the higher-level API going to expose
their choice of instrumentation to their users? With the current
Memprof.start API (before this patch), they would have to either
provide their own `start` function wrapping Memprof.start, or provide
a tuple of callbacks for to their users to pass to Memprof.start
themselves.

    val start : params -> unit
    (* or *)
    val callback : params ->
      ((allocation -> foo option) * (allocation -> bar option) * ... )

With an explicit record, it is easier for libraries to expose an
instrumentation choice (possibility parametrized over
user-provided settings):

    val tracker : params -> (foo, bar) Gc.Memprof.tracker

In addition, providing a record instead of optional parameters makes
it much easier to provide "default settings" (helper functions) that
instantiates the types `'minor` and `'ḿajor`, see for example
`simple_tracker` in this patch (which stores the same information for
the minor and major heap, and does not observe promotion), or to later
define checking predicates that can verify that a given choice of
callbacks is sensible (for example: providing a major-dealloc callback
but no promotion callback (dropping all tracked value on promotion) is
probably not a good idea).

### CODE REVIEW

**Comment by gasche** [comment]
(cc @stedolan, @jhjourdan)

File: stdlib/gc.mli
Comment (gasche) [inline stdlib/gc.mli]: I added a trailing `unit` parameter to this signature, in case we might want to evolve the API by adding extra optional parameters later. (If you don't have a non-labelled parameter after your optional parameters, things are broken.)

**Comment by gasche** [comment]
I considered also changing the C runtime to keep the tracker record, as a single generational root, instead of one root per callback as is currently done, to make the code slightly shorter/simpler (there isn't any performance impact). I am open to feedback from @jhjourdan on whether that would be a good idea.

**Comment by gasche** [comment]
I did the change to keep a single value for all the callbacks in a second commit, and I do find the result nicer.

**Comment by gasche** [comment]
I had forgotten to update the code of the tests at first (of course...). I have fixed this now, and this motivated to make slight changes to the tracker-building helper functions: there is now `alloc_tracker` which runs imperative code on each allocation and does not keep any metadata, and `alloc_dealloc_tracker` which records metadata on each allocation, and provides it back on deallocation.

Looking at the tests part of the diff is interesting because it shows how the API change impacts real code. (Although the tests are slightly artificial in that they use non-standard trackers.)

File: runtime/memprof.c
Comment (dra27) [inline runtime/memprof.c:58]: It'd be handy to have a reminder comment of the OCaml type this relates to (i.e. Gc.Memprof.tracker)

File: stdlib/gc.mli
Comment (dra27) [inline stdlib/gc.mli]: Possibly worth putting these in a module of their own? e.g. `Gc.Memprof.Trackers` or something?

**Review by dra27 — COMMENTED** [review]
Couple of comments - I agree that the code from the resulting API is much nicer.

One other thought - perhaps the `stop` function could return the tracker?

IIRC reducing the number of generational roots is always good - that benefit to the GC must surely outweigh the extra indirection to get the callback pointer on an already expensive call from C back to OCaml (i.e. for calling the callbacks). i.e. even if the OCaml API change is rejected, the C change feels good regardless.

File: runtime/memprof.c
Comment (gasche) [reply runtime/memprof.c:58]: Excellent point, I added a comment.

**Comment by gasche** [comment]
Re. the generational roots: we are talking about a constant-number reduction in domain-global roots, so I don't think that there would be any noticeable performance impact (in either direction), it's merely a matter of having sligthly {shorter, more readable} C code.

**Comment by stedolan** [comment]
I like the C code changes and I think the new API is good, and I'd be happy to merge this for those reasons alone.

However, I don't buy the idea that profiling libraries will expose `Gc.Memprof.tracker`s and let the user `start` them. Any use of Memprof I can think of involves updating some global state in the callbacks, which generally needs to be initialised right before Memprof is started. I don't see how this would be done reliably if the starting of Memprof is not under the control of the profiling library.

File: stdlib/gc.mli
Comment (jhjourdan) [inline stdlib/gc.mli:485]: Nitpick: it seems slightly weird to me that this functions takes one labelled parameter and one non-labelled parameter. If I had to choose, I would make all them labelled.

File: stdlib/gc.mli
Comment (jhjourdan) [inline stdlib/gc.mli:480]:
```suggestion
    (** Default callbacks simply return [None] or [()]. *)
```

**Review by jhjourdan — COMMENTED** [review]
I like the new API, and this nicely simplifies the code in the C runtime. I don't think this will have any noticeable performance impact in one way or another (come on, we are speaking about 1 vs. 5 generational roots!).

The only thing that I am not really convinced is the need for the basic trackers defined in `Gc.Memprof`. I imagine that the client libraries would have anyway enough code around their call to `Gc.Memprof.start` that writing all the fields is not an issue. Also, wrapping callbacks has some runtime cost (an indirect call), which, even if small, is probably useless.

Perhaps the goal is to allow for compatibility of client libraries with future versions of memprof, where new callback could be added. But then this should be documented and/or the record should be made private (or abstract) to make sure that the tracker type is only built using the constructors given by the library.

> However, I don't buy the idea that profiling libraries will expose Gc.Memprof.trackers and let the user start them. Any use of Memprof I can think of involves updating some global state in the callbacks, which generally needs to be initialised right before Memprof is started. I don't see how this would be done reliably if the starting of Memprof is not under the control of the profiling library.

I completely agree with that.

File: stdlib/gc.mli
Comment (gasche) [reply stdlib/gc.mli:485]: The numerical parameters are labelled because otherwise we cannot know what they mean from the type or the callsite. The tracker parameter doesn't need labelling for this reason, as it is clear what it means in both types and callsites.

**Comment by gasche** [comment]
> However, I don't buy the idea that profiling libraries will expose `Gc.Memprof.tracker`s and let the user `start` them. Any use of Memprof I can think of involves updating some global state in the callbacks, which generally needs to be initialised right before Memprof is started. I don't see how this would be done reliably if the starting of Memprof is not under the control of the profiling library.

I'm not sure the global state needs to be initialized "right before starting", as long as only the tracker callbacks will write to the state; it's likely to be some mutable statistics-aggregation structure. An API could be as follows:

```ocaml
(* library code *)
type state
type info
val report : out_channel -> state -> unit
val make : unit -> state * (info, info) tracker

(* user code *)
let () =
  let histo, tracker = Histogram.make () in
  Gc.create_alarm (fun () -> Histogram.report stderr histo);
  Gc.Memprof.start ~sampling_rate:0.01 tracker;
  at_exit (fun () -> Gc.Memprof.stop(); Histogram.report stderr histo)
```

This API makes it possible for users to reuse the same tracker/state across several Memprof-monitored program fragments, but that actually sounds like a reasonable thing to do in many scnarios.

On the other hand, the API also makes it possible to loose track of the association between the state and the tracker (closing over this state), and using the wrong state for a tracker or conversly. This is not very good. Maybe the tracker callbacks could all take some global state context as a first parameter.

```ocaml
(* gc.mli *)
type ('state, 'minor, 'major) = {
  alloc_minor: 'state -> allocation -> 'minor option;
  ...
}
val start : ... -> ('state, 'minor, 'major) tracker -> 'state -> unit

(* library *)
type state
type info
val tracker : (state, info, info) tracker
val make : unit -> state

(* user *)
let () =
  let histo = Histogram.make () in
  Gc.create_alarm (fun () -> Histogram.report stderr histo);
  Gc.Memprof.start ~sampling_rate:0.01 Histogram.tracker histo;
  at_exit (fun () -> Gc.Memprof.stop(); Histogram.report stderr histo)
```

Personally I think that these kind of first-order APIs are easier to explain and use flexibly than higher-order APIs that encapsulate everything (start/stop) in continuation-passing-style.

**Comment by gadmm** [comment]
I agree that the tracker might reasonably close over resources; doing some partial closure-conversion to move them outside of the tracker does not remove the concern. "Continuation-passing style" is unfortunately the current state of affairs for that, so I am not sure something is gained in the general case.

I have also thought that being able to query the Memprof settings in a record would be useful for composing them. (In that case you also want to get the sampling rate.) For instance in the case of allocation limits implemented with Memprof (as part of the logic of your program) you still want to be able to profile your program (without interfering with the program logic). On the other hand, it might be good enough if such libraries for allocation limits simply re-export the Memprof interface.

**Comment by gasche** [comment]
I don't agree with the description of the state-explicit API proposed in my previous message as "partial closure conversion". I am not taking out free variables from closures abstractly, I am giving an explicit/visible status to a specific piece of captured state that the user (or at least the library author) does have to think about in their design, namely the global instrumentation state. This state is something that is naturally part of the problem domain, so it is more natural to have it explicit; it is the fact of hiding it from the types that is a transformation, not the other way around.

That said, even though I am not fully satisfied with it, I still believe that the API proposed in the PR (so not the state-explicit one in the comment above) is an improvement over the current optional parameters, and there seems to be agreement over this. I think we should move to converge and approve/merge the PR to make progress.

The only remaining question is the idea of providing simpler tracker-building functions than just `null_tracker`. I thought it was an interesting opportunity given by the new API, and it does simplify the writing of the various tests, but I agree that they deviate from our design stance of "not building any user-friendliness in the runtime bits". Unless there are more comments on this, I am planning to remove them tomorrow.

File: stdlib/gc.mli
Comment (gasche) [reply stdlib/gc.mli]: (I ended up removing all of them except `null_tracker`.)

**Comment by gasche** [comment]
I removed all trackers except `null_tracker`, and I would be interested in getting approval for this PR (or clear recommendations for changes), which I think is an improvement over the current state, even if not perfect.

**Review by jhjourdan — APPROVED** [review]

**Comment by jhjourdan** [comment]
I gave you your approval :)

Yet another remark? Is there any reason for including the bootstrap in this PR? This is not needed, since `ocamlc` does not use Memprof.

**Comment by gasche** [comment]
I had to do a bootstrap during development because the system was screaming about the disappearance of the `caml_memprof_start_byt` primitive, used by the stdlib. Did I do something wrong?

**Comment by jhjourdan** [comment]
Ah, right!

**Comment by gasche** [comment]
(If we want this to be merged, we also need someone with commit rights to approve, for example @stedolan or @dra27 who looked at the PR already.)

File: stdlib/gc.mli
Comment (jhjourdan) [reply stdlib/gc.mli:485]: I don't care too strongly anyway, I just feel anesthetic to mix the two styles.

**Comment by gasche** [comment]
I do not have the feeling that the API proposed in my PR is perfect, but there seems to be a good consensus that it is better than the existing one. I'm just out of a not-so-fun rebase bootstrap, so I think that I will consider @jhjourdan's approval as valid enough, and merge if the CI goes green again.

For me in three months maybe: to bootstrap this PR one has to first remove `start` and `stop` from `stdlib/Gc.Memprof`, and recreate a dummy `caml_memprof_start_byt` function such as the following, because its presence is required by the bootstrap `stdlib.cmi`.

```c
CAMLprim value caml_memprof_start_byt(value* argv, int argn)
{
  return Val_unit;
}
```

**Comment by gadmm** [comment]
An objective reason why it is better is that with a record it is easy to quantify `'major` and `'minor` existentially to hide this implementation detail. The user who wants to manipulate different trackers  has to re-do this patch as boilerplate otherwise (I ran into that as part of my experiments).

How severe is the bootstrap issue? Can other people fall into it?

**Comment by gadmm** [comment]
I think what you wrote about the boostrap is the perfect sort of thing that should be included in the commit log instead of just having it as a comment on the proprietary development platform.

**Comment by dra27** [comment]
For the bootstrap, I think the “better” route is to have the new primitives committed on its own - when rebasing those are commits which need to be “edited” to redo the bootstrap

**Comment by gasche** [comment]
There were test failures corresponding to tests added in the statmeprof testsuite after I wrote the initial PR.

I'm not sure what to say about benchmarking: yes, everyone may run into a situation where the bootstrap is slightly tricky because they are removing a primitive mentioned in the bootstrapped files. My best recommendation would be to wait until that happens to you and then figure it out (at least this is what I did); there is some documentation in BOOTSTRAP.adoc (which I think we could/should move to boot/HACKING.adoc for consistency with other HACKING files).

(Regarding the commit/bootstrap pattern, the approach in the PR has the advantage of preserving bisectability (a bootstrap comes with the commit that requires the bootstrap), but it makes the change more painful to rebase. I think this was not the right choice but I don't know if I want to put energy into changing this -- all choices are annoying anyway.)

**Comment by gadmm** [comment]
I did some bootstrap before but yours seemed more complicated and you probably complain about it for a good reason.

I was wondering if this API improvement really needed the change in runtime, moreover the new runtime interface (saving the record) does not seem that natural to me in light of the changes I proposed at #9267 for instance. But, as you say, it might not be worth going back and now you ran out of steam. I think this is a good enough state for a nice improvement.

**Comment by dra27** [comment]
I don’t follow how you lose bisectability? All commits are buildable in both versions?

For what it’s worth (ie don’t change anything here!), retrospectively, the multicore guys would be having an easier time if we committed repeatable bootstraps...

**Comment by gasche** [comment]
The bootstrap path I performed replaces `caml_memprof_start` with another version of the same name but a different API. This means that some intermediate states of the build (the boot compiler linked with the new runtime) are incorrect. You are correct that this may not hurt bisectability, in paticular it may be the case that the testsuite (which does not test the boot compiler, iirc.) passes correctly.

Your point about helping Multicore and other big-rebases (eg. Bucklescript) is a good one, I will rework this PR to make it easy to bootstrap on replay/rebase.

**Comment by gasche** [comment]
After an hour of my time and three attempts, I managed to get a slightly better bootstrap story.

**Comment by stedolan** [comment]
> After an hour of my time and three attempts, I managed to get a slightly better bootstrap story.

For the record: thank you for this hour! Replaying elided bootstrap history in the alternate-timeline of multicore is not much fun.

**Pull request merged**
